modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
+              last = LOWER;
+            } else { // last == UN
+              result = false;
+              break;
+            }
+          }
+          else if _byte_isUpper(b) {
+            if last == UN {
+              last = UPPER;
+            } else { // last == UPPER || last == LOWER
+              result = false;
+              break;
+            }
+          } else {
+            // Uncased elements
+            last = UN;
+          }
+        }
+      }
+      return result;
+    }
+
+    proc toLower() /*: string*/ {
+      var result: string = this;
+      if result.isEmptyString() then return result;
+
+      for i in 0..#result.len {
+        const b = result.buff[i];
+        if _byte_isUpper(b) {
+          result.buff[i] = b + 0x20;
+        }
+      }
+      return result;
+    }
+
+    proc toUpper() /*: string*/ {
+      var result: string = this;
+      if result.isEmptyString() then return result;
+
+      for i in 0..#result.len {
+        const b = result.buff[i];
+        if _byte_isLower(b) {
+          result.buff[i] = b - 0x20;
+        }
+      }
+      return result;
+    }
+
+    proc toTitle() /*: string*/ {
+      var result: string = this;
+      if result.isEmptyString() then return result;
+
+      param UN = 0, LETTER = 1;
+      var last = UN;
+      for i in 0..#result.len {
+        const b = result.buff[i];
+        if _byte_isAlpha(b) {
+          if last == UN {
+            last = LETTER;
+            if _byte_isLower(b) {
+              result.buff[i] = b - 0x20;
+            }
+          } else { // last == LETTER
+            if _byte_isUpper(b) {
+              result.buff[i] = b + 0x20;
+            }
+          }
+        } else {
+          // Uncased elements
+          last = UN;
+        }
+      }
+      return result;
+    }
+
+    proc capitalize() /*: string*/ {
+      var result: string = this.toLower();
+      if result.isEmptyString() then return result;
+
+      var b = result.buff[0];
+      if _byte_isLower(b) { // Only change alpha
+        result.buff[0] = b - 0x20;
+      }
+      return result;
+    }
+
+  } // end record string
+
+
+  // We'd like this to be by ref, but doing so leads to an internal
+  // compiler error.  See
+  // $CHPL_HOME/test/types/records/sungeun/recordWithRefCopyFns.future
+  pragma "donor fn"
+  pragma "auto copy fn"
+  proc chpl__autoCopy(s: string) {
+    if debugStrings then
+      chpl_debug_string_print("in autoCopy()");
+
+    // This pragma may be unnecessary.
+    pragma "no auto destroy"
+    var ret: string;
+    const slen = s.len; // cache the remote copy of len
+    if slen != 0 {
+      if s.locale.id == chpl_nodeID {
+        if debugStrings then
+          chpl_debug_string_print("  local initCopy");
+        ret.buff = chpl_mem_alloc(s._size.safeCast(size_t),
+                                  CHPL_RT_MD_STR_COPY_DATA): bufferType;
+        memmove(ret.buff, s.buff, s.len.safeCast(size_t));
+        ret.buff[s.len] = 0;
```

Comment:
Or, at least, we could create unowned copy of strings pointing to C string literals.... that might require our string record have 3 states:
 - owned
 - unowned (but could point to another string which could be mutated)
 - unowned-constant (points to c string literal or equivalent)